torchqdynamics/odeint.py

```diff
@@ -1,80 +1,122 @@
+        return _odeint_main(*params)
+    elif gradient_algorithm == GradientAlgorithm.ADJOINT:
+        return _odeint_adjoint(*params)
+    else:
+        raise ValueError(f'Gradient algorithm {gradient_algorithm} not defined')
+
+
+def _odeint_main(qsolver, y0, t_save, exp_ops, save_states):
     if isinstance(qsolver.options, FixedStep):
-        return _fixed_odeint(qsolver, y0, tsave)
+        return _fixed_odeint(
+            qsolver, y0, t_save, qsolver.options.dt, exp_ops, save_states
+        )
     elif isinstance(qsolver.options, AdaptativeStep):
-        return _adaptive_odeint(qsolver, y0, tsave)
+        return _adaptive_odeint(qsolver, y0, t_save, exp_ops, save_states)
 
 
-def odeint_inplace(qsolver, y0, tsave):
+# for now we use *args and **kwargs for helper methods that are not implemented to ease the potential api
+# changes that could occur later. When a method is implemented the methods should take the same arguments
+# as all others
+def _odeint_inplace(*args, **kwargs):
     # TODO: Simple solution for now so torch does not store gradients. This
     #       is probably slower than a genuine in-place solver.
     with torch.no_grad():
-        return odeint_main(qsolver, y0, tsave)
+        return _odeint_main(*args, **kwargs)
 
 
-def odeint_adjoint(qsolver, y0, tsave, parameters):
+def _odeint_adjoint(*_args, **_kwargs):
     raise NotImplementedError
 
 
-def _adaptive_odeint(qsolver, y0, tsave):
+def _adaptive_odeint(*_args, **_kwargs):
     raise NotImplementedError
 
 
-def _fixed_odeint(qsolver, y0, tsave):
+def _fixed_odeint(qsolver, y0, t_save, dt, exp_ops, save_states):
     # initialize save tensor
-    ysave = torch.zeros(len(tsave), *y0.shape).to(y0)
+    y_save, exp_save = None, None
+    if save_states:
+        y_save = torch.zeros(len(t_save), *y0.shape).to(y0)
+
+    if len(exp_ops) > 0:
+        exp_save = torch.zeros(len(t_save), len(exp_ops)).to(
+            device=y0.get_device(), dtype=torch.float
+        )
+
+    # save first step
     save_counter = 0
-    if tsave[0] == 0.0:
-        ysave[0] = y0
+    if t_save[0] == 0.0:
+        if save_states:
+            y_save[0] = y0
+        for j, op in enumerate(exp_ops):
+            exp_save[save_counter, j] = torch.trace(op @ y0)
         save_counter += 1
 
-    # get qsolver fixed time step
-    dt = qsolver.options.dt
-
-    # run the ODE routine
-    t, y = 0.0, y0
-    while t < tsave[-1]:
+    # run the ode routine
+    y = y0
+    t, t_max = 0, max(t_save)
+    while t < t_max:
         # iterate solution
         y = qsolver.forward(t, dt, y)
-        t = t + dt
 
         # save solution
-        if t >= tsave[save_counter]:
-            ysave[save_counter] = y
+        if t >= t_save[save_counter]:
+            if save_states:
+                y_save[save_counter] = y
+
+            for j, op in enumerate(exp_ops):
+                exp_save[save_counter, j] = torch.trace(op @ y)
             save_counter += 1
+        t += dt
```

Comment:
If `t_max / dt` is not an integer, the solution will be iterated for a longer time than `t_max`. At some point I had put something like
```python
if t + dt > t_max:
   dt = t_max - t
```
at the beginning of the loop to avoid this problem. However, I agree that this is not ideal for time-independent problems for which it's nice to have `dt` constant.

I would argue that the best solution would be to update `dt` to `dt = t_max / math.ceil(t_max / dt)` somewhere in the code. 

Actually, even a stronger criterion requires that all values inside `t_save` are multiples of `dt`. Otherwise, the solution is saved to `y_save[i]` at a different time than `t_save[i]`.

Reply:
I think the solution is indeed to require that all values inside `t_save` are multiples of `dt`. However, if one just wants to see the global evolution and doesn't care about details (just plot Wigners for example) he wouldn't mind that `t_save` values are not perfectly aligned (if `dt` is small). IMO we should just make the doc clear to explain this, later. 
 And also I think we don't really care that the scheme is iterated for a tiny bit longer in case the final `t_save` value is not a multiple of `dt`.

Reply:
I don't really agree with this. It's quite important that we iterate for exactly the right time. If you're simulating a gate and iterate for 1% longer, you get 1% fidelity error, whereas your gate might otherwise have 1e-5 error.

Also, for the adjoint method, I need to run "smaller" odeints between every value of `t_save`. So it's quite important that all values of `t_save` are well aligned with `dt`. So, in my mind there are three solutions: (1) we update `dt` at the beginning of `_odeint_fixed`, (2) we enforce that all values of `t_save` are aligned with `dt` in `check_t_save`, or (3) we allow for the varying `dt` inside the loop, which is dynamically changed at the last time step in the loop.

Reply:
Maybe one solution would be, inside `check_t_save`, to:
(1) update `dt` such that `dt = t_max / math.ceil(t_max / dt)`
(2) update `t_save` such that `t_save[i] = dt * k_i` for some integer `k_i` (we just take the closest value)
And then:
(3) return the actual `t_save` to the user at the end, within the `result` instance.

Reply:
I agree with you, I think we should favor the simplest option (2) for now then (which we can do in another PR).

Reply:
Concerning
> It's quite important that we iterate for exactly the right time. If you're simulating a gate and iterate for 1% longer, you get 1% fidelity error, whereas your gate might otherwise have 1e-5 error.

This is very dependent indeed on the simulation you're running:
- quick qualitative simulation: you don't care about `t_save` being perfectly aligned
- quantitative simulation: you do care

Fixing the `dt` would fit both use cases BUT for testing and benchmarking numerical schemes + user experience I don't think you don't want some dark magic to update the `dt` in the background. Thus option (2) checking that the user chose compatible `dt` and `t_save` is the best choice I think.

Reply:
Yes I think option (2) is good. Plus it's probably what most people do naturally anyway.
So we can just add inside `check_t_save`:
```python
if torch.any(torch.remainder(t_save, dt) != 0.0):
    raise ValueError('Every value of argument `t_save` must be a multiple of the time step `dt`.')
```
I can do it in another PR, or we add it here, as you wish !

Reply:
Actually the above solution doesn't work. Can use this instead:
```python
if not torch.all(torch.isclose((tsave / dt).int().float(), tsave / dt)):
```